WFCORE-3416 User redirected with HTTP 301 instead of 302 in admin-onl…

TomasHofman opened the pull request:
…y mode

https://issues.jboss.org/browse/WFCORE-3416
https://issues.jboss.org/browse/JBEAP-10363

bstansberry commented on domain-http/interface/src/main/java/org/jboss/as/domain/http/server/ManagementRootConsoleRedirectHandler.java:
```diff
@@ -47,15 +46,8 @@ public void handleRequest(HttpServerExchange exchange) throws Exception {
-            redirect.append(consoleHandler.getContext()).append(consoleHandler.getDefaultPath());
-            exchange.getResponseHeaders().add(Headers.LOCATION, redirect.toString());
-            MOVED_PERMANENTLY.handleRequest(exchange);
+            consoleHandler.getHandler().handleRequest(exchange);
```
What happens with ConsoleMode.SLAVE_HC?

TomasHofman replied:
That behaves the same way as ADMIN_ONLY, only with different URL.

To make sure, I tested this way:
1) Added http management interface into host-slave.xml:
```
<http-interface security-realm="ManagementRealm">                   
    <http-upgrade enabled="true"/>                                  
    <socket interface="management" port="${jboss.management.http.port:9990}"/>                                                                     
</http-interface>
```
2) $ ./domain.sh --host-config=host-slave.xml ...
3) $ curl -s -D - http://123.0.0.1:9990/ 
HTTP/1.1 302 Found
Connection: keep-alive
Location: /consoleerror/noConsoleForSlaveDcError.html

bstansberry replied:
Sorry, I misread this and thought it was saying it was giving a 301, which is arguably correct since a slave does not provide a console in normal operation.

But, thinking about it harder I think the 302 is better. A slave is not a slave forever; it can be reconfigured manually and reloaded as a master, and once we do HA DC some slaves will be capable of dynamic switching from slave to master.

darranl replied:
Moving the status from 301 to 302 only impacts this initial redirect, once the browser receives the redirect the remainder of the caching for things like the console remain in place so the browser only requests those if required.  

Initially the server state was reasonably static in this area so the 301 worked well but with ongoing enhancements we have a variety of situations where this is no longer the case so having the root level redirect happen every time a user attempts to use the console is a good thing to ensure they are using it in relation to the current state of the server.

bstansberry commented on testsuite/manualmode/src/test/java/org/jboss/as/test/manualmode/adminonly/WebConsoleRedirectionTestCase.java:
```diff
@@ -0,0 +1,76 @@
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ *
+ */
```
Please use the ASL header in all new files in WildFly Core.

TomasHofman replied:
Will fix.

bstansberry commented on testsuite/manualmode/src/test/java/org/jboss/as/test/manualmode/adminonly/WebConsoleRedirectionTestCase.java:
```diff
@@ -0,0 +1,76 @@
+ *
+ */
+
+package org.jboss.as.test.manualmode.adminonly;
```
This should not need a manualmode test, which is more costly in terms of time vs one that uses an existing server and simply reloads it into / out of admin-only.

TomasHofman replied:
Will fix.

bstansberry commented on testsuite/manualmode/src/test/java/org/jboss/as/test/manualmode/adminonly/WebConsoleRedirectionTestCase.java:
```diff
@@ -0,0 +1,76 @@
+        final HttpURLConnection connection = getConnection();
+        connection.setRequestMethod(HttpGet.METHOD_NAME);
+        connection.connect();
+        assertEquals(HttpURLConnection.HTTP_MOVED_TEMP, connection.getResponseCode());
```
The fact this passes makes me suspect something is wrong. Without the change to ManagementRootConsoleRedirectHandler what response code is returned? A 404 or a 301?

If it's a 301 there's a separate issue that we'd need to sort out. There is no console possible in a simple WildFly Core server. The console module isn't even present. A redirect to a page that talks about admin-only is not valid.

If it's a 404, then we're good. The test of admin-only would be in WildFly full which is where we ship a console.

TomasHofman replied:
It returns 301 before the change. I tried starting the container in normal mode instead of admin-only, and that returns 404 - no console. I will look into why it gives different results.

So the test case would have to go into wildfly testsuite. Probably best if I create a separate WFLY issue then.

Thanks for feedback!

bstansberry replied:
Thanks. My guess is the logic that decides what ConsoleHandler to use is checking things in the wrong order.

TomasHofman replied:
Okey, so first the BaseHttpInterfaceAddStepHandler sets consoleEnabled attribute to true [1], because that's the default value for the attribute (console-enabled is undefined in the model).

After that, HttpManagementAddHandler selects either ConsoleMode.CONSOLE or ConsoleMode.ADMIN_ONLY [2], based on whether we are in admin-only or not.

If ADMIN_ONLY is selected, it will simply be used to create redirect handler with 302 (which however is not used in current code, hence this PR).

But, if ConsoleMode.CONSOLE was selected, handler creation will fail because of "ModuleNotFoundException: WFLYDMHTTP0007: No console module available with module name org.jboss.as.console.main" [3]. The console handler therefore stays null, which leads to 404 being returned [4] eventually.

Can we set console-enabled to false in default standalone.xml configuration file? I think it's generated in core-feature-pack module. Would that affect wildfly full somehow?

[1] https://github.com/wildfly/wildfly-core/blob/master/controller/src/main/java/org/jboss/as/controller/management/BaseHttpInterfaceAddStepHandler.java#L75
[2] https://github.com/wildfly/wildfly-core/blob/master/server/src/main/java/org/jboss/as/server/operations/HttpManagementAddHandler.java#L204
[3] https://github.com/wildfly/wildfly-core/blob/master/domain-http/interface/src/main/java/org/jboss/as/domain/http/server/ManagementHttpServer.java#L350
[4] https://github.com/wildfly/wildfly-core/blob/master/domain-http/interface/src/main/java/org/jboss/as/domain/http/server/ManagementRootConsoleRedirectHandler.java#L61

TomasHofman commented:
Test case: https://github.com/wildfly/wildfly/pull/10657

darranl commented:
Retest this please

bstansberry commented on core-feature-pack/src/main/resources/configuration/standalone/template.xml:
```diff
@@ -46,7 +46,7 @@
         </audit-log>
         <management-interfaces>
-            <http-interface security-realm="ManagementRealm">
+            <http-interface security-realm="ManagementRealm" console-enabled="false">
```
This is probably a good thing to do in general, but I'm a bit concerned changing this could affect someone in unexpected ways. (It doesn't affect full WildFly which has its own templates.) If we did this we'd want to change the host.xml and host-master.xml templates too.

The last commit at https://github.com/wildfly/wildfly-core/compare/master...bstansberry:pullRequest2934 results in admin-only behaving the same as normal when no console module is present.

darranl replied:
I agree, I think it is better if the fix is in the code rather than changing a template - an end user can always tweak the config effectively breaking the fix, if the fix is in the code it is always present.

darranl commented:
Just tested this locally, this does clear up the issue I describe in the Jira where after enabling SSL as the browser was previously passed a permanent redirect it uses a cached version of the console triggering cross origin issues when the console loads.

TomasHofman commented:
I pushed changes:
- the template change removed,
- added Brian's commit causing `DisabledConsoleHandler` to return 404, if console is not present,
- added test case verifying that in both modes the handler returns 404.

bstansberry commented on testsuite/standalone/src/test/java/org/jboss/as/test/integration/management/console/WebConsoleRedirectionTestCase.java:
```diff
@@ -0,0 +1,76 @@
+
+/**
+ * Since console is not loaded in core dist, in both modes the server is supposed to return 404.
+ * 
```
The PR is failing checkstyle due to trailing spaces on this line.

Otherwise. looks good.

TomasHofman replied:
Didn't notice the failure, fixed.
